### ISSUE TITLE
Use latest manifest timestamp for GC instead of Utc::now

### ISSUE THREAD

**Issue by agavra** [slatedb/slatedb#356]
**Is your feature request related to a problem? Please describe.**
After RFC-3 for TTL/Timestamps, operations on SlateDB include a timestamp and #325 makes it so that we can now access the most recent persisted timestamp from the manifest. Since garbage collection works in terms of `min_age`, using the timestamp in the last persisted manifest as the current time makes GC more predictable (and probably more reproducible).

**Describe the solution you'd like**
Read the manifest during GC and set the current time based on that.

**Describe alternatives you've considered**
We can stick with using wall clock time, but that could be risky if a machine is running with a bad clock or something like that.

**Additional context**

**Comment by agavra** [comment]
Ah I just realized that we use just the object store metadata and not the actual object store file contents (which makes sense, because we probably don't want to pull unnecessary data down) during GC.

This might cause some problems for when/if we want to implement fancier mechanisms such as TWCS (if all rows are expired, we can GC the entire SST). I wonder if there's something in-between we can do that's still efficient.

**Comment by criccomini** [comment]
This is making me wonder, long-term, what is the vision for where we use wall clock, where we use user-defined clock, and where we allow both? We use timing in a few places now (gc, ttl, snapshot, any others?), and I'm a little worried that, absent a clear vision, things could get really complicated later on...

Initially, I was thinking maybe we make everything use `Clock`, and then default that to wall-clock. I haven't thought through the implications of this if users implement weird clocks, though...

**Comment by agavra** [comment]
> Initially, I was thinking maybe we make everything use Clock, and then default that to wall-clock. I haven't thought through the implications of this if users implement weird clocks, though...

I think the requirement that clocks are monotonic really helps prevent users from implementing weird clocks and (IMO) makes it possible for us to use `Clock` as the defacto unit of time wherever possible. I think having multiple definitions of time really complicates a system (@hachikuji was reflecting about this decision in Kafka) and it also makes it more difficult to test (and I suspect more difficult to formally verify as well).

If we imagine time as only progressing globally when a manifest is committed, that feels like a really clean abstraction (and obviously locally time can progress between manifest commits). It also allows you to "rollback" to a specific point in time if we create time-aligned snapshots/checkpoints.

I'll give it some more thought, but I think we have a chance here with SlateDB to do something special 😄 (especially for the realtime/event-driven application space where tiem is such an important concept)

**Comment by criccomini** [comment]
Nice. This is along the lines of what I was thinking.

One note: technically, I think time progresses on manifest write _or_ SST WAL wrote, since those can be ahead of the manifest.

If we go down this path, we definitely need create_time in the SST and SsTableInfo structure. As you said, it'd require more reads, but I think that's OK. Should be pretty minimal sizes (just filter and index really).

**Comment by rodesai** [comment]
IMO we should not discard wallclock time outright. I think for some things you really do care about elapsed time in the real world rather than the time the data in the db is aligned to. Db time advances independent of wallclock time and that can cause problems if you're implementing some mechanism that hinges on how long things in the real world take like a lambda function execution, for example:
- When creating a checkpoint, you probably want to be able to create it with some lifetime that guarantees it will exist for some real-world duration, and potentially also that it will be cleaned up relatively soon. E.g. a lambda function that opens a reader just needs to create a checkpoint that's around for the duration of the lambda - it shouldn't matter if the db suddenly ingests a bunch of data that advances the clock quickly. On the other end of the spectrum, you should be able to get rid of the data after the checkpoint expires even if the db's clock hasn't advanced.
- Even for this max_age case I feel like it makes sense to use wallclock time because we're using it to ensure that SSTs stick around for some time so the writer has a chance to update the manifest before the last sst it wrote could be deleted by the GC. The writer needs some interval during which it can safely do WAL writes without refreshing the manifest. There's no way to guarantee that with db time.

Generally, I think for things that control correctness (like expiring data from the db based on its ttl) we want to use db time, but for "safety" (maybe not the right word) e.g. keeping data around for some time so the db can be rolled back after an upgrade, keeping a checkpoint around so a user can access it for some user-specified time, keeping WAL SSTs around so we don't break fencing it probably makes sense to use wallclock time.  

I think we need to model these types of time explicitly so it's clear what a given instant/interval refers to.

**Comment by agavra** [comment]
I feel like wall clock time has all the same problems as logical db-time, but without the guarantees that the entire system agrees on it. If there's a process where there's heavy load that causes a thread to pause or a network partition with retries it's possible that wall clock time progresses suddenly very quickly (though certainly not likely to be as far off as a db bootstrapping a ton of data with stream time). Similarly, wall clock can even go backwards in time in some situations. 

The point being here is that I think we should design the protocols to work with the constraints of logical time. To your examples:

- If you need any kind of lifetime guarantee for a checkpoint, you should use infinite retention and manual deletion. If you're running a lambda function, for example, it makes sense to clean up your checkpoints manually at the end of the lambda instead of hoping that it finishes within the time you're expecting it to run. Worse, if you're running tasks on spot instances or something like that you could be in for a lot of uncertainty in the time that it takes to complete your task.
- I think when it comes to the fencing protocol, this is exactly where I think it makes sense to design the protocols with the assumption that we have a logical clock. If we use wall clock, any number of things could go wrong (the active writer might stall for long enough that GC kicks in even with wall clock). If we, instead, design GC so that it uses the time from the latest manifest, that's not a problem because it'll always be behind the active WAL segments. That makes the interval during which you can safely write WALs infinite so long as no newer manifest is written (at which point the writer would be fenced anyway).

If anything, I feel like safety should be easier to guarantee with a perfectly synchronized clock (even if it's not a very granular clock).

**Comment by rodesai** [comment]
> though certainly not likely to be as far off as a db bootstrapping a ton of data with stream time

This is the crux of the problem - you can't set any reasonable bound across clients on how fast db-time progresses, whereas although wallclock time is not perfectly synchronized (though it's pretty close when you have true time available - e.g. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/compare-timestamps-with-clockbound.html) you can make some reasonable assumptions like your thread not being unscheduled for hours (although maybe not now that vm snapshots are a thing 😄). On the other hand db-time can easily suddenly advance by any arbitrary amount. db-time doesn't even need to represent time - it just has be to a monotonically increasing number.

This makes it impractical and hard to reason about for use cases where you want something to stick around for some period of real-world time like a reader running on a lambda or a spot instance. Lambda and spot instances both have no guarantees that the program will run to completion - they can be terminated in the middle of execution so there's no way to ensure an inifinite checkpoint will be cleaned up. The most practical approach I can think of is to use a checkpoint that won't be cleaned up before the program finishes running most of the time, which is easiest to express in wallclock time.

I'm less convinced about the wal fencing example. I feel like there may be some way to design the protocol to take advantage of db-time. But again, the fenced writer can't observe or make any assumptions about db time progression on the active writer between it's own WAL writes which makes it tricky.

> > EDIT: Another thought, the fact that SlateDB only has a single writer is what makes it uniquely possible to synchronize the clock with this mechanism and we don't have to rely on really complicated logical timestamps progressing across nodes because any operation where granularity of time is sensitive happens on the active node, which can have its own view of time that is ahead of the committed manifest time.

This is exactly why it's reasonable to use db-time to make decisions about data and queries, which are totally local to a given client, but makes it less suited to coordinate between processes.

**Comment by agavra** [comment]
I'm having a lot of thoughts but no conclusion yet, and I don't want to lose track of them when I go off to the ⛰️ for Thanksgiving:

1. **Design Space** I think it's a fair assumption that most use cases will have a clock that either is wall clock (default) or resembles wall clock in steady state (something like stream time). While other clocks ("absurd" clocks) should not break anything, I think it's acceptable for such apps to need to do additional work to get semantics they want.
2. **Defaults** The background processes in Slate are pretty modular, so we're really discussing default behaviors. If we go with db-time based expiry for checkpoints, someone could always implement a registry that tracks wall-clock based expiry and manually call delete on a checkpoint or vice versa (this "registry" can be as simple as having snapshot names with an expire_ts encoded in them).
3. **Bootstraps** If a bootstrap happens that bootstraps months of db-time data in a short period of wall-clock time I want to be able to clean up the junk that results from that bootstrap operation. I suspect bootstrapping/bulk loading use cases will be pretty common.
4. **Garbage Bloat** OTOH, given data is in S3, maybe it's not so bad if junk lives in S3 longer than it's supposed to but it's pretty bad if it gets cleaned up prematurely.
5. **Slow DB-Clocks** Along the lines of (4), the worst case is if wall clock time progresses faster than stream time. It's pretty bad if things get cleaned up when an app is "paused". What happens today if an active writer flushes some WALs, but doesn't yet flush a manifest, and then gets scaled to 0 and only scaled back up some long time after? Is it possible the WALs get cleaned up by a scheduled GC job?

-----

>  But again, the fenced writer can't observe or make any assumptions about db time progression on the active writer between it's own WAL writes which makes it tricky.

This took me a sec to wrap my head around. Is this what you're saying?

0. WAL configured to stay around for 1min
1. `W` writes WAL 1-4
2. `W` writes Manifest at db-time 5min, wall-clock 5m00s
3. `W'` writes WAL 5-8
4. `W'` writes Manifest at db-time 10min, wall-clock 5m10s
5. GC can now kick in and remove anything < 9min, so WAL 1-7 (for example) are cleaned
6. `W` successfully writes WAL-5 (problem)

Basically the problem in #352 is potentially exacerbated by situations where db-time progresses too quickly?

**Comment by agavra** [comment]
OK had time to mull this over during the break and I'm failing to see a way around using wall clock for some of the operations @rodesai mentioned. We probably will have to have some mix of operations where some use db-time, some use wall-clock time and perhaps some use both (so we can guarantee safety as with case 5 - slow db clocks - in my previous post)

**Comment by criccomini** [comment]
> where some use db-time, some use wall-clock time and perhaps some use both

This was where my head was at, too. I confess I haven't thought deeply about this subject yet, but it seems most intuitive to me. I think it would helpful to concretely list every location where clocks are used in the system and then think through which locations should use wall clock, which should use stream^H^H^H^H^H^HDB time, and which should use both.
